verus/lcf/src/kernel.rs

```diff
@@ -0,0 +1,1108 @@
+    pub open spec fn spec_subst(self, s: SpecSubst) -> (res: SpecTerm)
+        recommends
+            self.spec_complete_subst(s),
+    {
+        match self {
+            SpecTerm::Var(v) => {
+                let u_option = s.index(v);
+                SpecTerm::Const(u_option)
+            },
+            SpecTerm::Const(_) => self,
+        }
+    }
+}
+
+impl Term {
+    //proves equality for Term types
+    pub fn term_eq(&self, other: &Self) -> (res: bool)
+        ensures
+            res <==> self.deep_view() == other.deep_view(),
+    {
+        match (self, other) {
+            (Term::Const(s), Term::Const(t)) => Const::const_eq(&s, &t),
+            (Term::Var(u), Term::Var(v)) => u.eq(v),
+            _ => false,
+        }
+    }
+
+    //checks whether Term items are either Const or are Vars with a key in the Subst TmpStringHashMap
+    pub fn complete_subst(self, s: &Subst) -> (res: bool)
+        ensures
+            res <==> self.deep_view().spec_complete_subst(s.deep_view()),
+    {
+        match self {
+            Term::Var(v) => s.contains_key(v.as_str()),
+            Term::Const(_) => true,
+        }
+    }
+
+    //checks if term variant is a base type
+    pub fn concrete(self) -> (res: bool)
+        ensures
+            res <==> self.deep_view().spec_concrete(),
+    {
+        match self {
+            Term::Const(_) => true,
+            Term::Var(_) => false,
+        }
+    }
+
+    //performs a substitution to achieve a Const variant of Term
+    pub fn subst(self, s: &Subst) -> (res: Term)
+        requires
+            self.deep_view().spec_complete_subst(s.deep_view()),
+        ensures
+            res.deep_view().spec_concrete(),
+            res.deep_view() == self.deep_view().spec_subst(s.deep_view()),
+    {
+        match self {
+            Term::Var(v) => {
+                let u_option = s.get(v.as_str());
+                Term::Const(u_option.unwrap().clone())
+            },
+            Term::Const(_) => self,
+        }
+    }
+}
+
+//#[verifier::external_body] needed because the verifier does not support the equality relation we are trying to leverage
+#[verifier::external_body]
+//checks whether 2 Vec<Term>s are equal or not
+pub fn terms_eq(a: &Vec<Term>, b: &Vec<Term>) -> (res: bool)
+    ensures
+        res <==> a.deep_view() == b.deep_view(),
+{
+    a == b
+}
+
+pub enum Prop {
+    App(String, Vec<Term>),
+    Eq(Term, Term),
+}
+
+pub enum SpecProp {
+    App(Seq<char>, Seq<SpecTerm>),
+    Eq(SpecTerm, SpecTerm),
+}
+
+impl DeepView for Prop {
+    type V = SpecProp;
+
+    open spec fn deep_view(&self) -> Self::V {
+        match self {
+            Prop::App(s, v) => SpecProp::App(s.view(), v.deep_view()),
+            Prop::Eq(t, e) => SpecProp::Eq(t.deep_view(), e.deep_view()),
+        }
+    }
+}
+
+impl Clone for Prop {
+    //#[verifier::external_body] needed to satisfy postcondition on line 253.
```

Comment:
Resurrecting @mmcloughlin's comment: remove the line number reference here, since line numbers will always go out of date.